lib/spack/spack/cmd/logs.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import gzip
+import os
+import shutil
+import sys
+
+import spack.cmd
+import spack.util.compression as compression
+from spack.cmd.common import arguments
+
+description = "print out logs for packages"
+section = "basic"
+level = "long"
+
+
+def setup_parser(subparser):
+    arguments.add_common_arguments(subparser, ["spec"])
+
+
+def _dump_byte_stream_to_stdout(instream):
+    if sys.platform == "win32":
+        # https://docs.python.org/3/library/sys.html#sys.stdout unconditionally
+        # recommends using .buffer to write binary data to sys.stdout, but
+        # https://docs.python.org/3/library/io.html#io.TextIOBase.buffer claims
+        # it is not available in all implementations, so only try accessing it
+        # on Windows, where stdout.fileno() is not useful
+        try:
+            outstream = sys.stdout.buffer
+        except AttributeError:
+            raise spack.main.SpackCommandError(
+                "This command must be invoked in a"
+                " context where stdout is accessible"
+                f" as a binary stream (stdout = {type(sys.stdout)})"
+            )
+    else:
+        outstream = os.fdopen(sys.stdout.fileno(), "wb", closefd=False)
+
+    shutil.copyfileobj(instream, outstream)
+
+
+def dump_build_log(package):
+    with open(package.log_path, "rb") as f:
+        _dump_byte_stream_to_stdout(f)
+
+
+def _logs(cmdline_spec, concrete_spec):
+    if concrete_spec.installed:
+        log_path = concrete_spec.package.install_log_path
+    elif os.path.exists(concrete_spec.package.stage.path):
+        dump_build_log(concrete_spec.package)
+        return
+    else:
+        raise spack.main.SpackCommandError(f"{cmdline_spec} is not installed or staged")
+
+    if not os.path.exists(log_path):
+        raise spack.main.SpackCommandError(f"No logs are available for {cmdline_spec}")
+
+    compression_ext = compression.extension_from_file(log_path)
+    fstream = None
+    try:
+        if not compression_ext:
+            fstream = open(log_path, "rb")
+        elif compression_ext == "gz":
+            fstream = gzip.open(log_path, "rb")
+        else:
+            raise spack.main.SpackCommandError(
+                f"Unsupported storage format for {log_path}: {compression_ext}"
+            )
+
+        _dump_byte_stream_to_stdout(fstream)
+    finally:
```

Comment:
just use `with fstream` instead of rolling your own

Reply:
> Like I suggested before:

Can you point out where you explicitly suggested this? I just looked through every one of your comments and didn't see it. You do have examples that use a context-manager approach to `open`, but in all cases that I see, they are discussing other things (e.g. [here](https://github.com/spack/spack/pull/42202#discussion_r1466233128) your primary complaint is the use of `decompressor_for`).

That being said, I'm aware of using the context manager approach to opening files, but:

```
if ...:
  fstream = ...
else:
  fstream = ...
<------------ something can happen here, e.g. a signal
with fstream as f:
   ...
```

so I don't consider the context manager sufficient in this case.

Again, if you explicitly requested using the context manager approach in an earlier comment, then I should have explained my reasoning there, and I apologize for not doing so until now.

Reply:
https://github.com/spack/spack/pull/42202#discussion_r1466233128

Reply:
> #42202 (comment)

If this is meant to be the example of you being explicit, I can say that comment is the exact example I gave of you not being explicit

> What do you think about the code that runs between when gzip.open opens the file and before it returns

That's up to python's implementation of context managers, and `gzip`'s implementation of `.open` as a context manager: I can  trust `with gzip.open as f` the same way I trust `with open() as f`.

> if you're truly worried about this you should open the file in binary mode in a context manager before wrapping it in a gzip.GZipFile, but I don't think that is worth doing.

I addressed the concern I had in the way that was natural to me, this notion of `finally` closing a file being messy seems extremely odd to me: reading into the interface of `gzip` to understand that it can wrap an already open stream is perfectly fine, but I don't see how it's any simpler than realizing we need to `close` our open file.

Now that you've pointed it out, I can change it, but my point is that holding up the PR on something like this seems to be emphasizing the wrong thing.